### ISSUE TITLE
BalancedKMeans.cs: Wrong assumption in Learn when calculating Cluster.Proportions

### ISSUE THREAD

**Issue by jagbarcelo** [accord-net/framework#450]
Near the end of the **Learn(...)** method (line 230, **for** statement) you calculate the proportions of each cluster, suposing that **shouldStop = converged(centroids, newCentroids);** returned true because of an actual convergence. This is not always the case when you had set up MaxIterations beforehand.

**converged(...)** can return true when the limit has been reached but no valid solution has yet been found. In that case **Cluster.Proportions** values are wrong because they are not really calculated with the actual proportions of the current iteration, but rather with the assumption that everything ended successfully and they have all the same ratio.

**Comment by jagbarcelo** [comment]
Even though **private int[] compute(...)** (in KMeans.cs) is calculated the same way, it does not expose the same flaw. I've intentionally testing several cases with the regular K-Means and MaxIterations=1 and I haven't seen a single case in which Cluster.Proportions was wrongly calculated: When all Proportions were equal, the clusters were balanced. When there was any proportion with a lower value than the rest, that cluster had less items. When a value was higher, it had more ítems. Everything as expected.

However, there must be something wrong in the BalancedKMeans implementation, since always returns Cluster.Proportions with all values being equal, despite that the cluster calculation might be unbalanced.

**Comment by cesarsouza** [comment]
Hi @jagbarcelo,

When attempting to get balanced clusters it is possible to enforce that the labels of the training set will be balanced, but it is impossible to find centroids that will produce balanced clustering for all possible testing data. 

This means that one cannot use the output of .Learn() to obtain new balanced clusterings for new data. Instead, it is needed to look at the BalancedKMeans object itself to see which balanced cluster labels it found during training.

Instead of doing
```
var clusters = kmeans.Learn(observations);
int[] labels = clusters.Decide(observations);
```

Please do
```
var clusters = kmeans.Learn(observations);
int[] labels = kmeans.Labels; // get the labels attributed to training samples after the last call to .Learn()
```

Of course, this should have been reflected better in the documentation.

Regards,
Cesar

**Comment by cesarsouza** [comment]
Fixed in 3.6.0.
